### PR TITLE
feat: Add password & PUT endpoint to Users

### LINKED ISSUE #51: Implement Login Endpoint
As a business user I want to persist login credentials  so that valid users can access their own account information and have access to functionalities on the website.

Tasks:
- [x] Update Users to hold Account info [Kevin]
- [x] Update Users controller for saving Account info [Alec]
- [x] Create controller for getting Login info [Kevin, Alec]
- [x] Create the route for POST /login [3dxp]

Acceptance Criteria:

- When a POST request is made against the /v1/users endpoint with a JSON request body with a first name, last name, username, and Account info. A response with the location header set with the relative URL of the created user with a success status code is returned to the client.
- When a POST request is made against the /login endpoint, the login credential is authenticated and the if valid will return a login "token".
- When an invalid Login credentials are given an error code 401 is returned.

### LINKED ISSUE #87: Implement Login Authentication to POST Endpoints
As a account holder, I want my login to be safe and secure so that my information doesn't get stolen.

**Tasks:**
- [x] Create PUT /user/{userId} for User update [Kevin, Alec]
- [x] Add Authentication to POST post endpoints [Kevin]
- [x] Add Authentication to POST caption endpoints [Alec]
- [x] Add Authentication to POST _rating endpoints [3dxp]
- [x] Add Authentication middleware... [3dxp]

**Acceptance Criteria:**
- When a request is made against the POST endpoints for posts it is authenticated. 
- When an unauthenticated request is made against an endpoints, an error code 401 is returned.
- Unauthenticated Users can still view site content.

### PR DESCRIPTION
**Pull request by zxkevin**
Added new migrations to update Users table with password column
Seed file added to update pre-existing seeds with a password.
POST endpoint for Users now requires a password.
PUT endpoint for {userID} added for update users functionality.
Bcrypt is used to salt and hash password for secure password storage.

#87 #51

### CODE REVIEW

File: packages/hackmcx-api/src/db/seeds/2022-03-29_1757_update_demo_users.cjs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+function seed(knex) {
+    return knex('users').update('password', '$2a$12$A0HScvmVMgyRKcy/BYm/tuYi7IetjaYNpzU/gVHBmtHu3EtWUPshq');
```
Comment (zxkevin) [inline packages/hackmcx-api/src/db/seeds/2022-03-29_1757_update_demo_users.cjs:2]: "password123"

File: packages/hackmcx-api/src/controllers/users.controller.js [hunk 2]
```diff
@@ -22,38 +23,49 @@ export async function getUsersByUserId(req, res){
         })
 }
 
-export async function postUser(req, res){
+function userValidation(req){
     const emptyUsername = !req.body.username || /^\s*$/.test(req.body.username);
```
Comment (3dxp) [inline packages/hackmcx-api/src/controllers/users.controller.js:27]: I think at this point, checking for null or whitespace should be a helper function, with the regex compiled.

File: packages/hackmcx-api/src/controllers/users.controller.js [hunk 3]
```diff
@@ -22,38 +23,49 @@ export async function getUsersByUserId(req, res){
...
     const emptyFirstName = !req.body.firstname || /^\s*$/.test(req.body.firstname);
-    const emptyLastName = !req.body.lastname || /^\s*$/.test(req.body.lastname);    
-    const noGivenImageUrl = !req.body.imageUrl;
+    const emptyLastName = !req.body.lastname || /^\s*$/.test(req.body.lastname);
+    const emptyPassword = !req.body.password || /^\s*$/.test(req.body.password);  
     const notPicture = !req.body.imageUrl || !/(https?:\/\/.*\.(?:png|jpg|gif|svg))/i.test(req.body.imageUrl);
+    const noGivenImageUrl = !req.body.imageUrl;
 
     if(emptyUsername){
```
Comment (3dxp) [inline packages/hackmcx-api/src/controllers/users.controller.js:34]: I think these validation checks can be factored out, and maybe applied elsewhere. 

Eventually we might want to look at just adding some reasonable exception handling and breaking all this validation out of the controller.

File: packages/hackmcx-api/src/controllers/users.controller.js [hunk 4]
```diff
@@ -22,38 +23,49 @@ export async function getUsersByUserId(req, res){
...
+
+    if(validation != 1){
         res.statusCode = 400;
-        res.send({error: "Invalid image URL: Image url is invalid."});
-        return
+        res.send(validation);
+        return;
     }
 
+    const hashPass = await bcrypt.hash(req.body.password, 12);
```
Comment (3dxp) [inline packages/hackmcx-api/src/controllers/users.controller.js:63]: Why is this a `const`?

**Review by 3dxp — CHANGES_REQUESTED** [review]

File: packages/hackmcx-api/package.json [hunk 5]
```diff
@@ -9,11 +9,12 @@
     "api-seed": "node ./src/index.js --seed"
   },
   "dependencies": {
+    "bcrypt": "^5.0.1",
+    "cors": "^2.8.5",
```
Comment (3dxp) [inline packages/hackmcx-api/package.json:13]: It seems like the only thing changing here is the addition of bcrypt. We might want to consider removing the other changes, even it they do reorganize the imports in alphabetical order.

**Review by acampb65 — CHANGES_REQUESTED** [review]
I think preexisting seeds need to be modified. The update demo users seed works fine if rows already exist, but when running seed data on empty database MySQL gets mad:
```
code: 'ER_NO_DEFAULT_FOR_FIELD',
    errno: 1364,
    sqlState: 'HY000',
    sqlMessage: "Field 'password' doesn't have a default value",
    sql: "insert into `users` (`first_name`, `imageUrl`, `last_name`, `username`) values ('first_name1', 'https://upload.wikimedia.org/wikipedia/commons/thumb/2/20/MJd1-.svg/100px-MJd1-.svg.png', 'last_name1', 'username1'), ('first_name2', 'https://upload.wikimedia.org/wikipedia/commons/thumb/1/17/MJd2v1-.svg/100px-MJd2v1-.svg.png', 'last_name2', 'username2'), ('first_name3', 'https://upload.wikimedia.org/wikipedia/commons/thumb/0/04/MJd3e-.svg/100px-MJd3e-.svg.png', 'last_name3', 'username3')"
```

**Comment by acampb65** [comment]
PUT user responds with 201 even if username doesn't exist.

example:
```
curl -X PUT -i http://localhost:3003/api/v1/users/nonExistentUsername \
-H 'Content-Type: application/json' \
-d '{"username":"user","password":"pass","firstname":"first","lastname":"last"}'
       
HTTP/1.1 201 Created
X-Powered-By: Express
Access-Control-Allow-Origin: *
Location: /api/v1/users/undefined
Date: Wed, 30 Mar 2022 18:11:43 GMT
Connection: keep-alive
Keep-Alive: timeout=5
Content-Length: 0

```

**Review by 3dxp — APPROVED** [review]

**Review by acampb65 — APPROVED** [review]
LGTM

**Pull request merged**